--- a/docs/schemas/domaincommon.rng
+++ b/docs/schemas/domaincommon.rng
@@ -2912,6 +2912,11 @@
             </attribute>
           </optional>
         </group>
+        <group>
+          <attribute name="type">
+            <value>qubes</value>
+          </attribute>
+        </group>
       </choice>
     </element>
   </define>
--- a/src/conf/domain_conf.c
+++ b/src/conf/domain_conf.c
@@ -554,6 +554,7 @@ VIR_ENUM_IMPL(virDomainGraphics, VIR_DOM
               "vnc",
               "rdp",
               "desktop",
+              "qubes",
               "spice")
 
 VIR_ENUM_IMPL(virDomainGraphicsListen, VIR_DOMAIN_GRAPHICS_LISTEN_TYPE_LAST,
@@ -1400,6 +1401,9 @@ void virDomainGraphicsDefFree(virDomainG
         virDomainGraphicsAuthDefClear(&def->data.spice.auth);
         break;
 
+    case VIR_DOMAIN_GRAPHICS_TYPE_QUBES:
+        break;
+
     case VIR_DOMAIN_GRAPHICS_TYPE_LAST:
         break;
     }
--- a/src/libxl/libxl_conf.c
+++ b/src/libxl/libxl_conf.c
@@ -1329,6 +1329,9 @@ libxlMakeVfb(virPortAllocatorPtr graphic
             if (VIR_STRDUP(x_vfb->keymap, l_vfb->data.vnc.keymap) < 0)
                 return -1;
             break;
+        case  VIR_DOMAIN_GRAPHICS_TYPE_QUBES:
+            libxl_defbool_set(&x_vfb->vnc.enable, false);
+            libxl_defbool_set(&x_vfb->sdl.enable, false);
     }
 
     return 0;
@@ -1407,6 +1410,10 @@ libxlMakeBuildInfoVfb(virPortAllocatorPt
         unsigned short port;
         const char *listenAddr;
 
+        if (l_vfb->type == VIR_DOMAIN_GRAPHICS_TYPE_QUBES) {
+            libxl_defbool_set(&b_info->u.hvm.qubes_gui, true);
+        }
+
         if (l_vfb->type != VIR_DOMAIN_GRAPHICS_TYPE_SPICE)
             continue;
 
--- a/src/conf/domain_conf.h
+++ b/src/conf/domain_conf.h
@@ -1406,6 +1406,7 @@ typedef enum {
     VIR_DOMAIN_GRAPHICS_TYPE_VNC,
     VIR_DOMAIN_GRAPHICS_TYPE_RDP,
     VIR_DOMAIN_GRAPHICS_TYPE_DESKTOP,
+    VIR_DOMAIN_GRAPHICS_TYPE_QUBES,
     VIR_DOMAIN_GRAPHICS_TYPE_SPICE,
 
     VIR_DOMAIN_GRAPHICS_TYPE_LAST
